src/Microsoft.AspNet.Mvc.Abstractions/ModelBinding/ModelStateDictionary.cs

```diff
@@ -656,70 +653,204 @@ IEnumerator IEnumerable.GetEnumerator()
-                    {
-                        continue;
-                    }
+        public struct PrefixEnumerable : IEnumerable<KeyValuePair<string, ModelState>>
+        {
+            private readonly ModelStateDictionary _dictionary;
+            private readonly string _prefix;
+
+            public PrefixEnumerable(ModelStateDictionary dictionary, string prefix)
+            {
+                if (dictionary == null)
+                {
+                    throw new ArgumentNullException(nameof(dictionary));
                 }
 
-                // Everything is prefixed by the empty string
-                if (prefix.Length == 0)
+                if (prefix == null)
                 {
-                    yield return entry;
+                    throw new ArgumentNullException(nameof(prefix));
                 }
-                else
+
+                _dictionary = dictionary;
+                _prefix = prefix;
+            }
+
+            public PrefixEnumerator GetEnumerator()
+            {
+                return _dictionary == null ? new PrefixEnumerator() : new PrefixEnumerator(_dictionary, _prefix);
+            }
+
+            IEnumerator<KeyValuePair<string, ModelState>> IEnumerable<KeyValuePair<string, ModelState>>.GetEnumerator()
+            {
+                return GetEnumerator();
+            }
+
+            IEnumerator IEnumerable.GetEnumerator()
+            {
+                return GetEnumerator();
+            }
+        }
+
+        public struct PrefixEnumerator : IEnumerator<KeyValuePair<string, ModelState>>
+        {
+            private readonly ModelStateDictionary _dictionary;
+            private readonly string _prefix;
+
+            private bool _exactMatchUsed;
+            private IEnumerator<KeyValuePair<string, ModelState>> _enumerator;
+
+            public PrefixEnumerator(ModelStateDictionary dictionary, string prefix)
+            {
+                if (dictionary == null)
                 {
-                    var charAfterPrefix = key[prefix.Length];
-                    switch (charAfterPrefix)
+                    throw new ArgumentNullException(nameof(dictionary));
+                }
+
+                if (prefix == null)
+                {
+                    throw new ArgumentNullException(nameof(prefix));
+                }
+
+                _dictionary = dictionary;
+                _prefix = prefix;
+
+                _exactMatchUsed = false;
+                _enumerator = null;
+                Current = default(KeyValuePair<string, ModelState>);
+            }
+
+            public KeyValuePair<string, ModelState> Current { get; private set; }
+
+            object IEnumerator.Current
+            {
+                get
+                {
+                    return Current;
+                }
+            }
+
+            public void Dispose()
+            {
+            }
+
+            public bool MoveNext()
+            {
+                if (_dictionary == null)
```

Comment:
Won't need this if the ternary at [line 754](https://github.com/aspnet/Mvc/pull/3327/files#diff-a2b0a5845b6336b458622ec4dacfb762R754) goes away.